# Homework #

Homework #3

Goal: In this homework you will briefly tag a few documents for person references. You will then compare your tags to a set of tags produced by myself and possibly other people. For the purposes of the submission you need only submit the comparison information with the set I've tagged.


Step 1:

Tag data. You'll want to sign up at https://dataturks.com/ . There you will want to create an account and setup a new project for tagging.

Once you've created an account you'll want to create a new dataset. (+ button on the left hand side.)
Select the Document Annotation option under Text annotations.
    Provide a dataset name (you can use whatever)
    for the list of entities just input Person
    for the tagging instructions provide whatever you would like
    Then hit submit
    
You will then hit the "Upload raw data" button. Here, go into the homework data directory and upload the zip file which has all the documents inside of it.

You can then tag each document. Simply highlight the piece of text that you think is a person. When you've identified all of them for a single document hit "move to done". Use the guidelines contained in the next cell.

When you finish tagging all 5 documents, go into the project you've tagged and hit the options button in the top right. You will then hit "download". You will want to have the "complete items" and "json" options selected. Hit download and you will then have a json file for all of your documents.

(If this is confusing, please send me an email. I can try to come up with screen shots if necessary.)


Annotation guidelines:
    These are from the ACE Guidelines for Person:
    
3.1 Persons (PER) 
    Each distinct person or set of people mentioned in a document refers to an entity of type Person. For example, people may be specified by name (“John Smith”), occupation (“the butcher”), family relation (“dad”), pronoun (“he”), etc., or by some combination of these. Dead people and human remains are to be recorded as entities of type Person. So are fictional human characters appearing in movies, TV, books, plays, etc. 

There are a number of words that are ambiguous as to their referent. For example, nouns, which normally refer to animals or non-humans, can be used to describe people. If it is clear to the annotator that the noun refers to a person in a given context, it should be marked as a Person entity. 

Examples: 
He is [a real turkey]
[The political cat of the year]
She’s known as [the brain of the family] 

Step 2:

Look through the below code.

You will see that the compare annotations method needs to be implemented. It should report back the number of matches and non-matching annotations

Implement the appropriate code to calculate these numbers.

In [14]:
import json
from pprint import pprint


# method to read annotation file
def annotation_processor(annotation_file):
    annotation_array = []

    # here we need to be careful and process each line of the annotation file separately
    read_annotation = open(annotation_file)
    for line in read_annotation:
        data = json.loads(line)
        annotation_array.append(data)

    # here we return an array of the individual annotations
    return annotation_array
    
# calling the annotation processor function
annotation_processor('/Users/varsha/PythonWorkspace/homework_3/Annotated/annotated.json')


[{'content': 'According to Polk County Sheriff’s Office Chief Deputy Jonathan Blackmon, Hardin is a Rome native who worked with RomeCares and Floyd Sheriff’s officials to get the supplies to a disaster relief zone.\n\n“We originally took in the supplies for the victims of Hurricane Harvey in Texas, but when Irma hit closer to home we felt it would be more beneficial to send them to Florida,” Blackmon said.\n\nThe team that went to Florida with the supplies helped with relief efforts in the Moore Haven area as well.',
  'annotation': [{'label': ['Person'],
    'points': [{'start': 379, 'end': 386, 'text': 'Blackmon'}]},
   {'label': ['Person'],
    'points': [{'start': 293, 'end': 296, 'text': 'Irma'}]},
   {'label': ['Person'],
    'points': [{'start': 267, 'end': 272, 'text': 'Harvey'}]},
   {'label': ['Person'],
    'points': [{'start': 74, 'end': 79, 'text': 'Hardin'}]},
   {'label': ['Person'],
    'points': [{'start': 64, 'end': 71, 'text': 'Blackmon'}]},
   {'label': ['Person'],


In [15]:
annotation_processor('/Users/varsha/PythonWorkspace/homework_3/Annotated/new_annotated.json')



[{'content': 'According to Polk County Sheriff’s Office Chief Deputy Jonathan Blackmon, Hardin is a Rome native who worked with RomeCares and Floyd Sheriff’s officials to get the supplies to a disaster relief zone.\n\n“We originally took in the supplies for the victims of Hurricane Harvey in Texas, but when Irma hit closer to home we felt it would be more beneficial to send them to Florida,” Blackmon said.\n\nThe team that went to Florida with the supplies helped with relief efforts in the Moore Haven area as well.',
  'annotation': [{'label': ['Person'],
    'points': [{'start': 399, 'end': 402, 'text': 'team'}]},
   {'label': ['Person'],
    'points': [{'start': 378, 'end': 386, 'text': ' Blackmon'}]},
   {'label': ['Person'],
    'points': [{'start': 245, 'end': 252, 'text': ' victims'}]},
   {'label': ['Person'],
    'points': [{'start': 134, 'end': 140, 'text': 'Sheriff'}]},
   {'label': ['Person'],
    'points': [{'start': 86, 'end': 96, 'text': 'Rome native'}]},
   {'label': ['P

In [16]:
# here we will create two objects to store the reference annotations and your own annotations
reference_annotations = annotation_processor('/Users/varsha/PythonWorkspace/homework_3/Annotated/annotated.json')

# here I am just putting the same file in... if I do this I would expect a perfect match
my_annotations = annotation_processor('/Users/varsha/PythonWorkspace/homework_3/Annotated/new_annotated.json')



In [17]:
def compare_articles(reference_annotations, my_annotations):              
    matches = 0
    nonmatches = 0
    for articles in range(len(reference_annotations)):                          # Loops on the number of Articles  
        if reference_annotations[articles] == my_annotations[articles]:   # Matches on the entire Article
            matches += 1
        else:
            nonmatches += 1
    return matches, nonmatches

In [18]:

### Compare at Annotation Level
def compare_annotations_prelim(reference_annotations, personal_annotations):    # Loops on the Annotations themselves
    ref_points = [(reference_annotations[i]['points'][0]['start'],reference_annotations[i]['points'][0]['end']) for i in range(len(reference_annotations))]
    per_points = [(personal_annotations[i]['points'][0]['start'],personal_annotations[i]['points'][0]['end']) for i in range(len(personal_annotations))]
                                                                                # Finds the Start and End points for every Annotation in every Article
    num_matches = len(set(ref_points) & set(per_points))                        # Interection of two lists, to compare same Start and End points in both files
    num_non_matches = len(set(ref_points) | set(per_points)) - num_matches      # Union of the two lists, to compare Start and End points that only occur in one file
    return num_matches, num_non_matches

In [19]:

def compare_annotations(reference_annotation_array, personal_annotation_array):
    num_matches = 0
    num_nonmatches = 0
    for j in reference_annotation_array:                                        #
        for k in personal_annotation_array:
            if (j["content"] == k["content"]):                                  # Ensures the same articles are being compared
                temp_num_matches, temp_num_nonmatches = compare_annotations_prelim(j["annotation"], k["annotation"])  # 
                #print(temp_num_matches, temp_num_non_matches)
                num_matches += temp_num_matches                                 # Implement the sum of the temp_num_matches to the num_matches
                num_nonmatches += temp_num_nonmatches
    return num_matches, num_nonmatches

Step 3:
With the numbers now available, try to determine the Cohen's Kappa for this dataset. (You can calculate this by hand if you prefer). Assume that the Probability of random agreement is 0.3.
Note, with this the proportionate agreement would be the number of matches divided by the number of matches + the number of non-matches.
The rest of the formula should be as discussed in class.
Report the numbers you calculate.

In [20]:

def cohens_kappa(article_matches, article_nonmatches):                          # Function to calculate Cohen's Kappa
    observed_agreement = article_matches / (article_matches+article_nonmatches) # Calculates Observed Agreement 
    chance_agreement = 0.3                                                      # Pre-defined Chance Agreement
    kappa =  observed_agreement - chance_agreement/(1-chance_agreement)         # Cohen's Kappa Formula
    return observed_agreement, chance_agreement, kappa


In [21]:

### Results for Articles Level
result1a = (compare_articles(reference_annotations, my_annotations))    
print("Reference Vs Personal: ", result1a[0],"Article matches, ",result1a[1],"Article nonmatches")

result1b = (compare_articles(reference_annotations, reference_annotations))     # Results compared against itself   
print("Reference Vs Reference: ", result1b[0],"Article matches, ",result1b[1],"Article nonmatches") 
del(result1b)

result1c = (compare_articles(my_annotations, my_annotations))       # Results compared against itself 
print("Personal Vs Personal: ", result1c[0],"Article matches, ",result1c[1],"Article nonmatches") 
del(result1c)

article_matches = result1a[0]
article_nonmatches = result1a[1]
cohens_kappa_articles = cohens_kappa(article_matches,article_nonmatches)        # Calculates Cohen's Kappa for Result1a
print("cohen's kappa at Article level: ", round(cohens_kappa_articles[2],4))

### Results for Annotations Level
result2a = (compare_annotations(reference_annotations, my_annotations))
print("Reference Vs Personal: ", result2a[0],"Annotation matches, ",result2a[1],"Annotation nonmatches")

result2b = (compare_annotations(reference_annotations, reference_annotations))    
print("Reference Vs Reference: ", result2b[0],"Annotation matches, ",result2b[1],"Annotation nonmatches") 
del(result2b)

result2c = (compare_annotations(my_annotations, my_annotations))    
print("Personal Vs Personal: ", result2c[0],"Annotation matches, ",result2c[1],"Annotation nonmatches") 
del(result2c)

article_matches = result2a[0]
article_nonmatches = result2a[1]
cohens_kappa_annotations = cohens_kappa(article_matches,article_nonmatches)
print("cohen's kappa at Annotation level: ", round(cohens_kappa_annotations[2],4))


Reference Vs Personal:  0 Article matches,  5 Article nonmatches
Reference Vs Reference:  5 Article matches,  0 Article nonmatches
Personal Vs Personal:  5 Article matches,  0 Article nonmatches
cohen's kappa at Article level:  -0.4286
Reference Vs Personal:  16 Annotation matches,  44 Annotation nonmatches
Reference Vs Reference:  31 Annotation matches,  0 Annotation nonmatches
Personal Vs Personal:  45 Annotation matches,  0 Annotation nonmatches
cohen's kappa at Annotation level:  -0.1619


In [22]:
cohens_kappa(article_matches, article_nonmatches)

(0.26666666666666666, 0.3, -0.16190476190476194)

Step 4:

You will want to write a short report of 1 or 2 paragraphs. In it you should describe:
    What kind of differences there are between your annotations and the ones I have provided.
        Look through the different annotations and suggest where I might have been mistaken in identifying people
    Explain if you think tagging persons are difficult. Do you think the guidelines should be improved?
    A brief explanation of the Cohen's Kappa you calculated. Do you think it might be high or low? Does it report anything useful?

There is a clear distinction between the Reference and Personal Annotations.  
While the same 5 articles were reviewed, the number of annotations differed:
(Reference: 6+5+7+11+2 = 31 vs Personal: 7+6+11+8+3 = 35)

With different numbers of annotations, there should be some matched and non-matched annotations.
The first analysis compared Articles to Articles, with a total number out of 5.  
As expected, zero articles matched 100%.  

The second analysis compared the Start and End character positions, and returned 9 matches vs 48 non-matches.  
This analysis is more detailed and practical for real-world application.


After skimming both sets of articles, it appears the Reference focuses on first letter capitalized words (ie: Irma) 
while the Personal focuses on pronouns and groups of people (ie: we, team).
In addition, as the objective was to tag 'person references', 'Irma' might be referring to the hurricane rather than a person. 

It is clear from this assignment that two people can have very different interpetations of the same objective. 
The Cohen's Kappa scores were -0.42 and -0.27 for article level and annotation level, respectively.
It is difficult to interpret the Kappa Scores without a baseline as reference, but the higher the better. 
In my opinion, 0.27 does not seem like a very successful score.     


